### PR TITLE
Forces for bonded interactions for use with Langevin dynamics

### PR DESCRIPTION
**Pull request by mlund**
This adds force calculations to bonded interactions and, with harmonic as an example, shows how forces can be implemented for arbitrary bond types. These implementations should/could be done in a separate PR.

- [x] add `force` functor to `BondData`
- [x] add `Bonded::force()`
- [x] implement force calculation in `Harmonic` incl. unittest
- [ ] double-check correctness of `Harmonic::forceFunc`. Sign on force OK?

Contributes to #283

### CODE REVIEW

File: src/energy.cpp [hunk 1]
```diff
@@ -589,6 +589,47 @@ double Bonded::energy(Change &change) {
...
+        for (auto bond : bonds) {                                  // loop over all bonds in group
+            assert(bond->forceFunc != nullptr);                    // the force function must be implemented
+            auto bond_forces = bond->forceFunc(distance_function); // get forces on each atom in bond
+            assert(bond->index.size() == bond_forces.size());
+            int j = 0;
+            for (int index : bond->index) { // loop over atom index in bond (relative to group begin)
+                auto absolute_index = std::distance(spc.p.begin(), group.begin()) + index;
+                assert(absolute_index < forces.size());
+                forces[absolute_index] += bond_forces[j++]; // add to overall force
+            }
```
Comment (rc83) [inline src/energy.cpp:618]: If `bond->index` is a vector, I would prefer to use a single loop index variable `i` to address both `bond->index[i]` and `bond_forces[i]`. The non-loop variable `j` is confusing.

File: src/energy.cpp [hunk 2]
```diff
@@ -589,6 +589,47 @@ double Bonded::energy(Change &change) {
...
+    }
+
+    for (auto bond : inter) {                              // loop over inter-molecular bonds
+        assert(bond->forceFunc != nullptr);                // the force function must be implemented
+        auto bond_forces = bond->forceFunc(distance_function); // get forces on each atom in bond
+        assert(bond_forces.size() == bond->index.size());
+        int j = 0;
+        for (int absolute_index : bond->index) {        // loop over atom index in bond (absolute index)
+            forces[absolute_index] += bond_forces[j++]; // add to overall force
+        }
```
Comment (rc83) [inline src/energy.cpp:629]: If `bond->index` is a vector, I would prefer to use a single loop index variable `i` to address both `bond->index[i]` and `bond_forces[i]`. The non-loop variable `j` is confusing.

File: src/energy.cpp (hunk 1 above)
Comment (mlund) [reply src/energy.cpp]: fixed

File: src/energy.cpp (hunk 2 above)
Comment (mlund) [reply src/energy.cpp]: fixed

**Pull request merged**